OperatorReduce - prevent multiple terminal events

davidmoten opened the pull request:
As per discussion in #4242, if an operator maps an `onNext` emission to an `onError` emission downstream then it needs be defensive about an `onCompleted` being sent from upstream even if upstream has been unsubscribed.

Includes a unit test that failed on the original code.

Also renamed test class to match tested class.

davidmoten committed:
OperatorReduce - prevent multiple terminal events

vanniktech commented on src/main/java/rx/internal/operators/OnSubscribeReduce.java:
```diff
@@ -77,19 +79,26 @@ public void onNext(T t) {
-            actual.onError(e);
+            if (!done) {
+                done = true;
+                actual.onError(e);
```
Isn't here `RxJavaHooks.onError` missing?

davidmoten replied:
yep, fixing

davidmoten commented:
- Updated with `RxJavaHooks.onError` 
- checked `done` at start of `onNext` 
- added unit tests including coverage of calling `.reduce` on an empty stream
